### PR TITLE
Define API and split up vanilla renderers

### PR DESCRIPTION
**Pull request by edgarmueller**
This PR mostly involves moving around types and function declarations as well as providing a common API interface to JSONForms. 
It also features some refactorings of the more complex vanillla renderers that is a result of trying to make use of the API, although it perhaps should have been made a PR itself.
This is WIP, as the materila renderers are probably broken and no final API yet has been defined.

### CODE REVIEW

File: packages/core/src/core.ts [hunk 1]
```diff
@@ -120,14 +120,4 @@ export class JsonForms {
...
- * Annotation for registering a class as JSONForms service.
- * @param config
- * @constructor
- */
-// Disable rule because it is used as an decorator
-// tslint:disable:variable-name
-export const JsonFormsServiceElement = () => (cls: JsonFormsServiceConstructable) => {
-  JsonForms.jsonFormsServices.push(cls);
-};
 // tslint:enable:variable-name
```
Comment (eneufeld) [inline packages/core/src/core.ts]: This should be removed, as the disablement was removed

File: packages/core/src/generators/index.ts [hunk 2]
```diff
@@ -1,2 +1,16 @@
...
+
 export * from './schema-gen';
 export * from './ui-schema-gen';
+
+import { generateJsonSchema } from './schema-gen';
+import { generateDefaultUISchema } from './ui-schema-gen';
+
+const Generate: {
+  jsonSchema(instance: Object, options?: any): JsonSchema;
+  uiSchema(jsonSchema: JsonSchema, layoutType?: string, prefix?: string): UISchemaElement } = {
```
Comment (eneufeld) [inline packages/core/src/generators/index.ts]: weird formatting

File: packages/core/src/generators/index.ts [hunk 3]
```diff
@@ -1,2 +1,16 @@
...
+
+  jsonSchema: generateJsonSchema,
+  uiSchema: generateDefaultUISchema,
+};
+export { Generate };
```
Comment (eneufeld) [inline packages/core/src/generators/index.ts]: shouldn't we export this as the default export, too?

File: packages/default/src/additional/array/index.ts [hunk 4]
```diff
@@ -0,0 +1,10 @@
+import {
+  registerStartupRenderer,
+} from '@jsonforms/core';
+import ArrayControlRenderer from './ArrayControlRenderer';
+import { arrayTester } from './tester';
```
Comment (eneufeld) [inline packages/default/src/additional/array/index.ts]: shouldn't we also export the renderer to allow to extend it?

File: packages/default/src/additional/categorization/index.ts [hunk 5]
```diff
@@ -0,0 +1,8 @@
+import { registerStartupRenderer } from '@jsonforms/core';
+import CategorizationRenderer from './CategorizationRenderer';
+import { categorizationTester } from './tester';
+
```
Comment (eneufeld) [inline packages/default/src/additional/categorization/index.ts]: Shouldn't we export all the components and the renderer to allow an easy extension?

**Comment by edgarmueller** [comment]
@eneufeld I've updated the PR

**Pull request merged**